### PR TITLE
ch4/shm: fix accumulate computation with pairtype

### PR DESCRIPTION
**Pull request by minsii**

### CODE REVIEW

File: src/mpid/ch4/src/ch4_impl.h [hunk 1]
```diff
@@ -1052,15 +1052,16 @@ static inline int MPIDI_CH4U_destroy_shm_segment(MPI_Aint shm_segment_len,
...
+ * src_packed must be set to 1.*/
 #undef FUNCNAME
 #define FUNCNAME MPIDI_CH4U_compute_acc_op
 #undef FCNAME
 #define FCNAME MPL_QUOTE(FUNCNAME)
 MPL_STATIC_INLINE_PREFIX int MPIDI_CH4U_compute_acc_op(void *source_buf, int source_count,
                                                        MPI_Datatype source_dtp, void *target_buf,
                                                        int target_count, MPI_Datatype target_dtp,
-                                                       MPI_Op acc_op)
+                                                       MPI_Op acc_op, int src_packed)
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4_impl.h]: Let's use `bool` and `true/false` for newly introduced variables.

Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: Good catch. Will fix.

Comment (hajimefu) [reply src/mpid/ch4/src/ch4_impl.h]: Or... define an enum
```C
{
     MPIDI_SRC_PACKED,
     MPIDI_SRC_UNPACKED,
}
```
then pass these values. That way we can avoid awkward comments like
```C
fun(..., 0 /* src_packed */);
```

Comment (raffenet) [reply src/mpid/ch4/src/ch4_impl.h]: 👍 I think enum is the generally preferred style for these types of arguments for the reason you show. Less interpretation for the reader.

Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: I am not sure if `UNPACKED` is suitable here, as the buffer is not "unpacked" from another buffer. The parameter is simply a TRUE/FALSE flag.

Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: Similer code I have seen in ch4 is 
```
enum {
    MPIDI_CH4R_NETMOD = 0,
    MPIDI_CH4R_SHM = 1
};
```
But this defines two types, not a true/false flag.

Comment (minsii) [reply src/mpid/ch4/src/ch4_impl.h]: Maybe `PACKED, PLAIN`.

**Comment by minsii** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by minsii** [comment]
All tests have been passed except the ch4/ofi ones. I believe they are irrelevant to this PR. Also locally checked that `rma/acc_pairtype_shm` fails with both ch4 and ch3 if not including the fix.

**Comment by raffenet** [comment]
How about `DEFAULT` instead of `PLAIN`? This looks good otherwise. If you agree with the name change, I can make it locally while signing off.

**Comment by minsii** [comment]
DEFAULT is OK too. Thanks.

Min

Sent from my iPhone

On Dec 20, 2018, at 7:52 AM, Ken Raffenetti <notifications@github.com<mailto:notifications@github.com>> wrote:


How about DEFAULT instead of PLAIN? This looks good otherwise. If you agree with the name change, I can make it locally while signing off.

**Comment by raffenet** [comment]
Re-testing with my name changes included.
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by minsii** [comment]
I will look into the failure `get_flushlocal_shm` in ch3/tcp. The others should be irrelevant.

**Comment by minsii** [comment]
@raffenet checked that `get_flushlocal_shm` also fails with master code. Maybe we can merge this PR and deal with the issue separately ?

**Comment by minsii** [comment]
Actually I just noticed that it is a bug in the test program. Added the test fix in this PR.
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by minsii** [comment]
@raffenet I think this PR is ready to merge. What do you think ?

**Pull request merged**